implementations/rust/ockam/ockam_core/src/transport/metadata.rs

```diff
@@ -0,0 +1,65 @@
+use crate::compat::string::{String, ToString};
+use crate::errcode::{Kind, Origin};
+use crate::transport::{LocalInfoTransportType, TRANSPORT_IDENTIFIER};
+use crate::{AddressMetadata, Error, Result, TransportType};
+
+/// Transport Metadata used for Terminal Address
+pub struct TransportMetadata {
+    transport_type: TransportType,
+}
+
+impl TransportMetadata {
+    /// Transport type
+    pub fn transport_type(&self) -> TransportType {
+        self.transport_type
+    }
+}
+
+impl TransportMetadata {
+    #[track_caller]
+    fn error_type_id() -> Error {
+        Error::new(
+            Origin::Identity,
+            Kind::Invalid,
+            "invalid metadata identifier for transport type",
+        )
+    }
+
+    #[track_caller]
+    fn error_format() -> Error {
+        Error::new(
+            Origin::Identity,
+            Kind::Invalid,
+            "invalid format for metadata identifier for transport type",
+        )
+    }
+
+    /// Get the transport type
+    pub fn from_terminal_address_metadata(terminal: &AddressMetadata) -> Result<Self> {
```

Comment:
The amount of dynamic typing we do here with putting untyped values in generic messages then decoding them in places we expect them makes me wonder if we could have better types. 

I guess that would require adding a type parameter for `IncomingAccessControl<M: Secured + Transported>` so that it works hand in hand with a `Worker { type Message = M }` and still allows us to check identifier + transport locality at the ABAC level, because we eventually pass around a `SecuredMessage<LocalMessage>`. 

I know that what I'm writing is really fuzzy at this stage but this kind of dynamic typing makes me think that there might be some better solutions in a language like Rust.